[5.6] Copy-propagation enables lexical-lifetimes.

nate-chandler opened the pull request:
Commits needed to enable lexical lifetimes when copy propagation is enabled.

nate-chandler committed:
[LoopRotate] Don't phi forwarded guaranteed value.

A guaranteed value produced by a begin_borrow can't be both used as
an operand of an ownership forwarding-instruction and also reborrowed by
being used as a phi argument.

Avoid that by stopping rotation when encountering a header block
containing an ownership-forwarding instruction whose forwarded ownership
kind is guaranteed; such a rotation may result in using both the
original guaranteed value and the resulting guaranteed value as phi
arguments.

nate-chandler committed:
[Test] Add two lexical lifetime verification tests.

Verify that lexical-lifetimes and copy-propagation interact correctly to
keep objects named by variables alive in a couple interesting cases:
unsafe pointers, synchronization point calls.

nate-chandler committed:
[CopyPropagation] Note ShrinkBorrowScope changes.

Use the result returned from shrinkgBorrowScope to update the changed
flag.

nate-chandler committed:
[Gardening] Merged two conditions.

nate-chandler committed:
[Gardening] Updated comment.

nate-chandler committed:
[OwnershipUtils] Extracted utility for common use.

Promoted isRedundantLexicalBeginBorrow function to OwnershipUtils so
that it can be used in more than just SemanticARCOpts.

nate-chandler committed:
[CanonicalizeInst] Remove redundant lexical bbis.

Previously, CanonicalizeInstruction::eliminateSimpleBorrows bailed out
when encountering any any [lexical] begin_borrow.  While generally such
borrow scopes must be preserved, they may be removed if they are
redundant (when the borrowed value is guaranteed by another means: an
outer lexical borrow scope or a guaranteed function argument).  Here,
removing such redundant lexical borrow scopes is enabled.

atrick committed:
Fix OperandOwnership for AutoDiff builtins.

AutoDiffAllocateSubcontext and AutoDiffProjectTopLevelSubcontext
return RawPointer. They cannot be ForwardingBorrows. This was
triggering an assert in ForwardingOperand's constructor.

nate-chandler committed:
Enabling copy propagation enables lexical lifetimes.

The effect of passing -enable-copy-propagation is both to enable the
CopyPropagation pass to shorten object lifetimes and also to enable
lexical lifetimes to ensure that object lifetimes aren't shortened while
a variable is still in scope and used.

Add a new flag, -enable-lexical-borrow-scopes=true to override
-enable-copy-propagation's effect (setting it to ::ExperimentalLate) on
SILOptions::LexicalLifetimes that sets it to ::Early even in the face of
-enable-copy-propagation.  The old flag -disable-lexical-lifetimes is
renamed to -enable-lexical-borrow-scopes=false but continues to set that
option to ::Off even when -enable-copy-propagation is passed.

nate-chandler committed:
[SIL] Reenable destroy hoisting under lexical lifetimes.

Until https://github.com/apple/swift/pull/40392 lands, run destroy
hoisting with lexical lifetimes.

nate-chandler committed:
[SILOptimizer] Add flag to print pre-OSSA lowering.

The new Xllvm flag prints the module to stdout just before the OSSA
lowering pass runs.

atrick committed:
SemanticARCOpts fix - remove deadEndBlocks checks

SemanticARCOptVisitor::performGuaranteedCopyValueOptimization was
converting this SIL

    %borrow = begin_borrow %copiedValue
    %copy = copy_value %borrow
    %borrowCopy = begin_borrow %copy
    end_borrow %borrow
    end_borrow %borrowCopy
    destroy_value %copy
    // something something
    unreachable

into

    %borrow = begin_borrow %copiedValue
    %innerBorrow = begin_borrow %borrow
    end_borrow %borrow
    end_borrow %innerBorrow
    // something something
    unreachable

Dead-end blocks are simply irrelevant for this
optimization. Unfortunately, there were multiple layers of attempted
workarounds that were hiding the real problem, except in rare cases.

Thanks Nate Chandler for reducing the test.

nate-chandler committed:
[AllocStackHoisting] New dealloc_stack's locs are cleanups.

AllocStackHoisting creates new dealloc_stack instructions.  The old
dealloc_stack instructions' locations' kinds are ::CleanupKind  The new
instructions' locations need to be of that same kind too in order not to
trigger the sil verifier failure

    Basic block contains a non-contiguous lexical scope at -Onone

when building at Onone.

nate-chandler commented:
@swift-ci please test

nate-chandler committed:
Disable lexical-lifetimes with copy-propagation.

If -disable-copy-propagation is passed, just emit lexical diagnostic
markers but do not enable lexical lifetimes.

nate-chandler committed:
[NFC] Renamed LexicalLifetimesOption cases.

The cases' new names more accurately reflect what behavior occurs when
SILOptions::LexicalLifetimes is assigned that case.

nate-chandler commented:
@swift-ci please clean test windows platform

atrick committed:
Test case for SemanticARCOpts fix - remove deadEndBlocks checks

atrick committed:
Workaround OSSA dead borrow scope bugs.

More bugs related to dead-end blocks and OSSA lifetimes that aren't
well formed because of that.

Independently, I'm working on prohibiting ill-formed OSSA even in the
presence of dead-end blocks. That makes these workarounds unnecessary,
but we urgently need a narrow fix here.

atrick commented:
We'll want to add at least one more once the PR is merged:
https://github.com/apple/swift/pull/40533

which also contains the test cases for the last commit here 😊

nate-chandler commented:
@atrick Added! Along with https://github.com/apple/swift/pull/40519 .

atrick commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -1449,21 +1449,88 @@ static bool ParseSILArgs(SILOptions &Opts, ArgList &Args,
+    } else {
+      enableLexicalLifetimesFlag = true;
+    }
+  }
```
It appears that, for frontend options, the -enable vs. -disable form is much easier to handle.

Sorry if I indicated otherwise. I typically use `-enable-xxx=true|false` for LLVM options

No urgent need to rewrite this, unless we make have build settings start passing them...

Also for future reference. We should have some simple utility that takes two complementary boolean flags and checks that they don't conflict rather than writing custom logic

nate-chandler replied:
These flags (which match https://github.com/apple/swift/pull/40473#pullrequestreview-826968897) are such that `-enable-lexical-lifetimes` has the same effect as `-enable-lexical-lifetimes=true`, so flags that are currently being passed will continue to work.

nate-chandler committed:
Fixed flag checking for lexical borrow scopes.

Previously the following pairs were prohibited incorrectly:
- -enable-lexical-borrow-scopes=true, -enable-experimental-move-only
- -enable-lexical-lifetimes=true, -enable-experimental-move-only
and the following pairs were allowed incorrectly:
- -enable-lexical-borrow-scopes=false, -enable-experimental-move-only
- -enable-lexical-lifetimes=false, -enable-experimental-move-only
Here, that's fixed.  The first two pairs are allowed and the second two
pairs prohibited.

nate-chandler committed:
[SILOpt] Tweaked error message.

nate-chandler committed:
Error on bad copy propagation flag combo.

Diagnose passing both -enable-copy-propagation and
-disable-copy-propagation.

nate-chandler committed:
[NFC] Used SILOption field for copy propagation.

Replaced the quad-state (of state which one was illegal) of two booleans
(EnableCopyPropagation and DisableCopyPropagation) with an enum.

nate-chandler committed:
[NFC] [SILOpt] Explicitly specified flag defaults.

In preparation to change the default values, specify explicitly what
they are.

nate-chandler committed:
Base LexicalLifetimes dflt on SILOpts::CopyProp.

Previously, the default value for SILOptions::LexicalLifetimes was based
on a copy propagation behavior (which can then be overridden by the
flags for lexical lifetimems) only when the copy propagation was
explicitly specified.  Instead, set base the default value for this
option (SILOptions::LexicalLifetimes) on the effective copy propagation
behavior (i.e. SILOptions::CopyPropagation) regardless of whether an
explicit behavior has been specified.

Doing so will ensure that the desired behavior occurs as the default
behavior for copy-propagation changes, but for now this change is NFC.

nate-chandler committed:
Replaced -disable-copy-prop with multi-var.

Replaced the -disable-copy-propagation flag with
-enable-copy-propagation=false where the latter is a new multi-var
-enable-copy-propagation= which can take one of three values:
- true
- requested-passes-only
- false

nate-chandler committed:
[Test] Specified copy-prop and lex-lt behavior.

In preparation for changing the default, explicitly specify the behavior
of all tests that are affected by the choice of behavior for lexical
lifetimes and copy-propagation.

nate-chandler committed:
[SILOpt] Make lex-lt options optional.

Use boolOrDefault to distinguish betweeen options that are unset and
those that are set (to some value).  Then, only override the default
values in SILOptions when the options are specified.

nate-chandler committed:
[Gardening] Added omitted word.

nate-chandler committed:
[CopyPropagation] Cleanup after borrow shrinking.

nate-chandler committed:
[ShrinkBorrowScope] Hoist over copies of borrow.

nate-chandler committed:
[OwnershipUtils] Repurpose utility.

Replaced findInnerTransitiveGuaranteeedUsesOfBorrowedValue with
findExtendedUsesOfSimpleBorrowedValue.  Starting from a borrowed value,
it finds all extended (i.e., seeing through copies) uses of the borrow
and its projections within the simple (i.e. without considering
reborrowing) borrow scope.

nate-chandler committed:
[Gardening] Improved variable name.

nate-chandler committed:
[ShrinkBorrowScope] Return modified copy_values.

So that CopyPropagation and other clients can react accordingly, pass
back a list of copy_value instructions that were rewritten by
ShrinkBorrowScope.  In CopyPropagation, add each modified copy to the
copy worklist.

nate-chandler committed:
[NFC] Demoted variable to function scope.

The blocksWithReachedTops variable is only ever used while finding
barriers and in a small helper function called at that time.  Make it
local to that function and make that helper a lambda in that function.

nate-chandler committed:
[Gardening] Doc'd field.

nate-chandler committed:
[Gardening] Improved variable name.

nate-chandler committed:
[Gardening] Improved variable name.

nate-chandler committed:
[ShrinkBorrowScope] Avoid spurious mutation.

If an end_borrow cannot be hoisted, do not rewrite it.

nate-chandler committed:
[Test] Stopped passing wrong flag.

Previously, ShrinkBorrowScope was only run alongside
CanonicalizeBorrowScope under the flag -canonical-ossa-rewrite-borrows.
That is no longer so, so remove that flag.  Update a test case to undo
the canonicalization which CanonicalizeBorrowScope was previously doing.

nate-chandler committed:
[Test] Added swift test for ShrinkBorrowScope.

The new file tests the -O pipeline up to just before OSSA lowering.  It
will be used for verifying the effects of ShrinkBorrowScope on swift
code directly, to complement the tests done on SIL by just running a
single pass.

nate-chandler committed:
[Test] Added more end_borrow hoisting tests.

nate-chandler committed:
[Test] Adapted tests to CopyPropagation changes.

nate-chandler committed:
[Gardening] Doc'd field.

nate-chandler committed:
[ShrinkBorrowScope] Enable querying hoistability.

nate-chandler committed:
[ShrinkBorrowScope] Checked before adding preds.

Previously, after determining that a block was dead, all its
predecessors were added to the worklist.  That is a problem, with the
current algorithm, if any of the predecessors have a terminator over
which the borrow scope cannot be shrunk.  Here, before adding a dead
block's predecessors to the worklist of blocks through which to shrink,
ensure that all those predecessors have terminators over which the scope
can be shrunk.  Finally, when beginning to process a block, if it
doesn't have a starting instruction, first hoist over thet terminator.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test source compatibility

atrick reviewed (approved):
Thanks for gathering all the dependent commits here.
